core/lib/parameter_hunter.js

```diff
@@ -13,110 +13,230 @@
+   *     that remaining commas and colons are either delimiters, or wrapped
+   *     within quotes to not be recognized as such.
+   *   * A do-while loop crawls paramString to write keys to a keys array and
+   *     values to a values array.
+   *   * Start by parsing the first key. Determine the type of wrapping quote,
+   *     if any.
+   *   * By knowing the open wrapper, we know that the next quote of that kind
+   *     (if the key is wrapped in quotes), HAS to be the close wrapper.
+   *     Similarly, if the key is unwrapped, we know the next colon HAS to be
+   *     the delimiter between key and value.
+   *   * Save the key to the keys array.
+   *   * Next, search for a value. It will either be the next block wrapped in
+   *     quotes, or a string of alphanumerics, decimal points, or minus signs.
+   *   * Save the value to the values array.
+   *   * The do-while loop truncates the paramString value while parsing. Its
+   *     condition for completion is when the paramString is whittled down to an
+   *     empty string.
+   *   * After the keys and values arrays are built, a for loop iterates through
+   *     them to build the final paramStringWellFormed string.
+   *   * No quote substitution had been done prior to this loop. In this loop,
+   *     all keys are ensured to be wrapped in double-quotes. String values are
+   *     also ensured to be wrapped in double-quotes.
+   *   * Unescape escaped unicodes except for double-quotes. Everything beside
+   *     double-quotes will be wrapped in double-quotes without need for escape.
+   *   * Return paramStringWellFormed.
+   *
+   * @param {string} pString
+   * @returns {string} paramStringWellFormed
+   */
   function paramToJson(pString) {
-    var paramStringWellFormed = '';
-    var paramStringTmp;
-    var colonPos;
-    var delimitPos;
-    var quotePos;
-    var paramString = pString;
-
+    var colonPos = -1;
+    var keys = [];
+    var paramString = pString; // to not reassign param
+    var paramStringWellFormed;
+    var quotePos = -1;
+    var regex;
+    var values = [];
+    var wrapper;
+
+    //replace all escaped double-quotes with escaped unicode
+    paramString = paramString.replace(/\\"/g, '\\u0022');
+
+    //replace all escaped single-quotes with escaped unicode
+    paramString = paramString.replace(/\\'/g, '\\u0027');
+
+    //replace all escaped colons with escaped unicode
+    paramString = paramString.replace(/\\:/g, '\\u0058');
+
+    //with escaped chars out of the way, crawl through paramString looking for
+    //keys and values
     do {
 
-      //if param key is wrapped in single quotes, replace with double quotes.
-      paramString = paramString.replace(/(^\s*[\{|\,]\s*)'([^']+)'(\s*\:)/, '$1"$2"$3');
+      //check if searching for a key
+      if (paramString[0] === '{' || paramString[0] === ',') {
+        paramString = paramString.substring([1], paramString.length).trim();
```

Comment:
I found another instance of using an array instead of a number for `substring()`'s first argument. Must have missed that last time around - sorry!